### ISSUE TITLE
Feature request: RMSProp without momentum variables

### ISSUE THREAD

**Issue by netheril96** [tensorflow/tensorflow#13546]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS High Sierra
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: v1.3.0-24-g658866597 1.3.0
- **Python version**: 3.6.3
- **Bazel version (if compiling from source)**: 0.6.1
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**:

### Describe the problem

The current `RMSPropOptimizer` always allocates momentum variables, even though the default (and probably 99% of people) never make use of it. In fact, if one wishes to combine momentum and adaptive gradient descent, he/she will most likely instantiate an `AdamOptimizer` instead. The extra variables waste precious GPU/CPU memory as well as disk space (when saved as checkpoints) while providing minimal utility. 

Suggestion: introduce a new version of `ApplyRMSProp` operations that doesn't use momentum variables at all, and dynamically choose which implementation to use in `RMSPropOptimizer` constructor depending on whether the `momentum` argument is constant zero.

Alternative solution: change the current `ApplyRMSProp` operations so that it doesn't use momentum variables, and direct the minority users who currently need momentum with RMSProp to Adam instead.

**Comment by facaiy** [comment]
If I understand you correctly, do you mean that intermediate variable `mom` is wasteful when `momentum = 0`?

https://github.com/tensorflow/tensorflow/blob/624bcfe409601910951789325f0b97f520c0b1ee/tensorflow/python/training/rmsprop.py#L32-L36

Sounds reasonable, at least for me. I prefer to the first solution, which seems harmless and feasible: 
+ stop allocating resources if `momentum = 0`:
   https://github.com/tensorflow/tensorflow/blob/624bcfe409601910951789325f0b97f520c0b1ee/tensorflow/python/training/rmsprop.py#L103
+ adjust interface and c++ kernel implementations.
   https://github.com/tensorflow/tensorflow/blob/624bcfe409601910951789325f0b97f520c0b1ee/tensorflow/core/kernels/training_ops.cc#L352-L353

The work seems hard but not difficult :-)

**Comment by netheril96** [comment]
@facaiy Yes that is exactly what I meant.

**Comment by tatatodd** [comment]
@jhseu Do you have ideas on who might be able to evaluate / work on this?

**Comment by jhseu** [comment]
I'd favor not changing it. Is the extra memory and disk usage a big issue in practice? The additional memory usage would be minimal compared to activations. Users who care about memory/disk usage could copy the optimizer and remove the momentum part.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by jhseu** [comment]
While I'm not opposed to this in general if there's an easy way of doing it, I am somewhat opposed to adding a new Op to save a slot unless there's a lot of demand for it. Closing this out for now.
